emacs/config.org
#+title: Doom Emacs Configuration
#+AUTHOR: James Patrick
#+SETUPFILE: ./dist/setup.org
#+TODO: TRAIL_RUN(t) REFACTOR_NEEDED(r) BROKEN(b) |
#+PROPERTY: header-args :mkdirp yes

This Emacs configuration is written in the style of [[https://en.wikipedia.org/wiki/Literate_programming][Literate Programming]] paradigm & using Hlissner's awesome [[https://github.com/doomemacs/doomemacs][Doom Emacs]] framework. I'll detail the "how" and "why" as I get a better figure out what the hell I'm doing. My current configuration is the result of trying to solidify a collection of outdated work around, snippets cribbed from Stack Overflow, ideas from folks smarter than I am, then held together with string & bubble gum. Please don't hesitate to tell me my elisp-fu sucks or if there's a better way to do anything. If you have any questions please let me know.

* Document Structure
** Doom Emacs
I'm utilizing Hlissner's [[https://github.com/doomemacs/doomemacs][Doom Emacs]] frameworks. Doom Emac's configuration is divided into 150 modules, & fair number of convenience functions. Rather than detailing stuff about I'll say to go read [[https://github.com/doomemacs/doomemacs/blob/master/docs/faq.org][FAQs]] & [[https://github.com/doomemacs/doomemacs/blob/f9664ae058d67b8d97cb8a9c40744fefc3e5479f/docs/getting_started.org][Getting Started]]. Also the [[https://github.com/doomemacs/doomemacs/blob/master/docs/index.org][Index]] if you want to be well informed.

Configuration is located in the ~$DOOMDIR~ directory; I'll be using =$XDG_HOME_DIR/doom= / =~/.config/doom=

** Tangle
This org-mode page works by using [[https://orgmode.org/manual/Extracting-Source-Code.html][tangles]]. Doom's literate config module adds any ~#+begin_src~ blocks into the =config.el= file[fn:1]. If you want to tangle to a specific file you pass in a file name.

#+begin_src org :tangle no
#+begin_src emacs-lisp :tangle "example.el"
#+end_src

or to disable it tangling for a code block set the =:tangle= property in code block header to =no=.

#+begin_src org :tangle no
#+begin_src orgmode :tangle no
#+end_src

*** No-Web Reference
This document also uses [[https://orgmode.org/manual/Noweb-Reference-Syntax.html][noweb reference syntax]] to try and DRY this doc out a little. For example

#+begin_example :tangle no
#+NAME: initialization
#+BEGIN_SRC emacs-lisp
  (setq sentence "Never a foot too far, even.")
#+END_SRC

#+BEGIN_SRC emacs-lisp :noweb yes
  <<initialization>>
  (reverse sentence)
#+END_SRC
#+end_example

will produce

#+begin_example
#+BEGIN_SRC emacs-lisp :noweb yes
  (setq sentence "Never a foot too far, even.")
  (reverse sentence)
#+END_SRC
#+end_example

** Files tree
There's a fair number of files located in ~$DOOMDIR~.  Some are configuration, some are additional tooling not use directly by emacs.

#+begin_src bash :tangle no :exports result :results output verbatim
tree $DOOMDIR -L 2
#+end_src

#+RESULTS:
#+begin_example
.
├── backup_init.el
├── config.d
│   ├── +keybinds.el
│   ├── +org.el
│   └── +ui.el
├── config.el
├── config.org
├── custom.el
├── dist
│   ├── bundle.js
│   ├── code.css
│   ├── custom.css
│   ├── setup.org
│   └── tufte.css
├── init.el
├── makefile
├── modules
│   ├── lang
│   └── tools
├── packages.el
└── snippets
    └── org-mode

8 directories, 17 files
#+end_example

*** Emacs Configuration Files
Doom has a lot of configuration has a lot of flexibility & a large number of doing things incorrectly. At its most basic there are 3 files create and used by default for your ~$DOOMDIR~.

+ =init.el= :: Enabled Doom Modules
+ =packages.el= :: Packages to fetch from MELPA, GitHub, etc.
+ =config.el= :: General Emacs configuration

In addition to these there exists the following special directories
+ =modules/= :: Custom modules. See [[https://github.com/doomemacs/doomemacs/blob/master/docs/getting_started.org#writing-your-own-modules][Writing your own modules]] for details. I recommend opting for this on the sooner side.
+ =autoload/= :: Functions instantiated very early.
+ =snippets/[major mode]/*= :: These are [[https://github.com/joaotavora/yasnippet][YASnippets]]. These are manage by this document. See [[*Snippets][Snippets]] for more.

*I've also added a non-standard ~$DOOMDIR/config.d~ directory. Any =.el= file added here will get automatically loaded as part of the =config.el= file.* See [[*Loading =$DOOMDIR/config.d=][Loading $DOOMDIR/config.d]] for the implementation.

Currently, this includes[fn:4]
- =+org.el= :: Additional configuration for Orgmode. Considering this is one of the primary modes I use emacs in there's a fair bit of configuration there.
- =+keybinds.el= ::  A relatively new and not fully utilized file. The plan is to use this as a centralized location for all custom key binds.
- =+ui.el= :: General Purpose UI changes.

We also have an additional backup_init.el file. This is used to stub in the =init.el= config with literate
*** Non-Doom Files
- =makefile= :: You know what makes a terrible alternative to things like [[https://www.gnu.org/software/stow/][stow]], [[https://github.com/yadm-dev/yadm][yadm]], or [[https://github.com/twpayne/chezmoi][chezmoi]]? Know what I'm using anyways?[fn:2]
- =dist/= :: These are files used by the output of [[https://github.com/ox-tufte/ox-tufte][ox-tufte]].[fn:3] TODO Move this to assets
- =.assets/= :: Images, CSS, Javascript, etc.
- =config.html= :: This org file exported from [[https://github.com/ox-tufte/ox-tufte][ox-tufte]]. This is not versioned.

**  =.el= file prefixing

There [[https://nullprogram.com/blog/2016/12/22/][minor but non-zero start time benefits]] for using lexical bindings comments. All files created should start with

#+name:lexical-binding
#+BEGIN_SRC emacs-lisp :tangle no
;;; -*- lexical-binding: t; -*-
#+END_SRC

Anytime we sync or save our =config.org= file we are going to clobber any tangled files. So lets remind ourselves about this foot gun.

#+name:modification-warning
#+begin_src emacs-lisp :tangle no
;;; This file is generated from config.org's tangles. Any modification will be
;;; clobbered on sync or save, make modifications to the config.org instead.
;;; Or don't. I'm a sign not a cop.
#+end_src

*** Applying to all files
We need to apply this to all the following files.

#+begin_src bash :tangle no :exports result :results output verbatim
FOO=${DOOMDIR:-${HOME}/.config/doom} find ${FOO}* -iname '*.el' | sed "s/^${FOO}//"
#+end_src

#+RESULTS:
#+begin_example
backup_init.el
config.d/+keybinds.el
config.d/+org.el
config.d/+ui.el
config.el
custom.el
init.el
modules/tools/llm/keybinds.el
modules/tools/llm/packages.el
modules/tools/llm/config.el
modules/lang/mermaid/packages.el
modules/lang/mermaid/package.el
modules/lang/mermaid/config.el
modules/lang/mermaid/doctor.el
packages.el
#+end_example

Since we've used a named these blocks we can use [[*No-Web Reference][No-Web Reference]] for all the files. Lets define a new section

#+name:elisp-header
#+begin_src emacs-lisp :tangle no :noweb no-export
<<lexical-binding>>
<<modification-warning>>
#+end_src

=config.el=
#+begin_src emacs-lisp :tangle "config.el" :noweb no-export :comments no
<<elisp-header>>
#+END_SRC

=init.el=
#+begin_src emacs-lisp :tangle "init.el" :noweb no-export :comments no
<<elisp-header>>
#+END_SRC

=packages.el=
#+begin_src emacs-lisp :tangle "packages.el" :noweb no-export :comments no
<<elisp-header>>
#+END_SRC

Then all of the files in =config.d=

=config.d/+ui.el=
#+begin_src emacs-lisp :tangle "config.d/+ui.el" :noweb no-export :comments no
<<elisp-header>>
#+END_SRC

=config.d/+keybinds.el=
#+begin_src emacs-lisp :tangle "config.d/+keybinds.el" :noweb no-export :comments no
<<elisp-header>>
#+END_SRC

=config.d/+org.el=
#+begin_src emacs-lisp :tangle "config.d/+org.el" :noweb no-export :comments no
<<elisp-header>>
#+END_SRC

** Loading =$DOOMDIR/config.d=
We want to load any =.el= file in the =config.d= directory, so lets use the following snippet.

#+begin_src emacs-lisp :tangle "config.el" :noweb no-export :comments no
(mapc (lambda (file)
        (load! file nil t))
      (directory-files (concat doom-user-dir "config.d") t "\\.el\\'"))
#+END_SRC

* Doom Modules
Doom's modules are enabled via the =init.el= file. I've parted this file out using noweb. If a module needs some tweak, or there some comments some part, it will be located there. This is based off a =init.example.el= pulled from [[https://github.com/doomemacs/doomemacs/blob/e6c755305358412a71a990fc2cf592c629edde1e/static/init.example.el][e6c7553]].


#+name: init.el
#+attr_html: :collapsed t
#+begin_src emacs-lisp :tangle "init.el" :noweb no-export :comments no
;; This file controls what Doom modules are enabled and what order they load
;; in. Remember to run 'doom sync' after modifying it!

;; NOTE Press 'SPC h d h' (or 'C-h d h' for non-vim users) to access Doom's
;;      documentation. There you'll find a link to Doom's Module Index where all
;;      of our modules are listed, including what flags they support.

;; NOTE Move your cursor over a module's name (or its flags) and press 'K' (or
;;      'C-c c k' for non-vim users) to view its documentation. This works on
;;
;;      Alternatively, press 'gd' (or 'C-c c d') on a module to browse its
;;      directory (for easy access to its source code).

(doom! :input
       <<doom-input>>

       :completion
       <<doom-completion>>

       :ui
       <<doom-ui>>

       :editor
       <<doom-editor>>

       :emacs
       <<doom-emacs>>

       :term
       <<doom-term>>

       :checkers
       <<doom-checker>>

       :tools
       <<doom-tools>>

       :os
       (:if (featurep :system 'macos) macos)  ; improve compatibility with macOS
       ;;tty               ; improve the terminal Emacs experience

       :lang
       <<doom-lang>>

       :email
       <<doom-email>>

       :app
       <<doom-app>>

       :config
       <<doom-config>>
)
#+end_src

** Input
https://github.com/doomemacs/doomemacs/tree/master/modules/input

This section is un-used by me but still worth keeping for reference

#+name:doom-input
#+begin_src emacs-lisp :tangle no
;;bidi              ; (tfel ot) thgir etirw uoy gnipleh
;;chinese
;;japanese
;;layout            ; auie,ctsrnm is the superior home row
#+end_src

** Completion
https://github.com/doomemacs/doomemacs/tree/master/modules/completion

#+name:doom-completion
#+begin_src emacs-lisp :tangle no
;;(company +childframe)           ; the ultimate code completion backend
(corfu +icons +orderless)  ; complete with cap(f), cape and a flying feather!
;;helm              ; the *other* search engine for love and life
;;ido               ; the other *other* search engine...
;;ivy               ; a search engine for love and life
(vertico +icons)          ; the search engine of the future
#+end_src

This is an example of something that should have auto-complete.

** UI
https://github.com/doomemacs/doomemacs/tree/master/modules/ui

#+name:doom-ui
#+begin_src emacs-lisp :tangle no
;;deft              ; notational velocity for Emacs
doom              ; what makes DOOM look the way it does
doom-dashboard    ; a nifty splash screen for Emacs
doom-quit         ; DOOM quit-message prompts when you quit Emacs
(emoji +unicode)  ; 🙂
hl-todo           ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
;;indent-guides     ; highlighted indent columns
;;ligatures         ; ligatures and symbols to make your code pretty again
;;minimap           ; show a map of the code on the side
modeline          ; snazzy, Atom-inspired modeline, plus API
nav-flash         ; blink cursor line after big motions
;;neotree           ; a project drawer, like NERDTree for vim
ophints           ; highlight the region an operation acts on
(popup +defaults)   ; tame sudden yet inevitable temporary windows
;;smooth-scroll     ; So smooth you won't believe it's not butter
tabs              ; a tab bar for Emacs
(treemacs +lsp)          ; a project drawer, like neotree but cooler
;;unicode           ; extended unicode support for various languages
(vc-gutter +pretty) ; vcs diff in the fringe
vi-tilde-fringe   ; fringe tildes to mark beyond EOB
;;window-select     ; visually switch windows
workspaces        ; tab emulation, persistence & separate workspaces
zen               ; distraction-free coding or writing
#+end_src

*** Fancy Welcome Image
Rather than the ascii doom logo. Lets use the Doom logo off the [[https://www.reddit.com/r/Doom/comments/1eu9cwy/a_look_at_all_the_doom_box_art_over_the_years/][original box art]]. This is enabled by the =doom-dashboard= module.

#+CAPTION:Doom Welcome Icon
[[./.resources/doom.png]]

#+begin_src emacs-lisp :tangle "config.d/+ui.el" :noweb no-export :comments no
(setq fancy-splash-image (concat doom-user-dir "./.resources/doom.png"))
#+end_src

** Editor
https://github.com/doomemacs/doomemacs/tree/master/modules/editor

#+name:doom-editor
#+begin_src emacs-lisp :tangle no
(evil +everywhere); come to the dark side, we have cookies
file-templates    ; auto-snippets for empty files
fold              ; (nigh) universal code folding
(format +onsave)  ; automated prettiness
;;god               ; run Emacs commands without modifier keys
;;lispy             ; vim for lisp, for people who don't like vim
;;multiple-cursors  ; editing in many places at once
;;objed             ; text object editing for the innocent
;;parinfer          ; turn lisp into python, sort of
rotate-text       ; cycle region at point between text candidates
snippets          ; my elves. They type so I don't have to
(whitespace +guess +trim)  ; a butler for your whitespace
;;word-wrap         ; soft wrapping with language-aware indent
#+end_src

*** Custom Formatter
Doom uses [[https://github.com/lassik/emacs-format-all-the-code][emacs-format-all-the-code]] for its formatting. It's a great package, but I don't always love the default option its decided on. Right now I'm only going to update it to use
At this time I'm only going to update =sql-mode= to use [[https://github.com/darold/pgFormatter][pgformatter]]. It's supported, but instead used by default.

#+begin_src emacs-lisp
(setq-hook! 'sql-mode-hook +format-with 'pgformatter)
#+end_src

It's also worth adding =nogrouping=0= to your =$XDG_CONFIG_HOME/pg_format/pg_format.conf= file. Everything should be wrapped in a transaction block. If you're wanting fall back points use checkpoints.

** Emacs
https://github.com/doomemacs/doomemacs/tree/master/modules/emacs

#+name:doom-emacs
#+begin_src emacs-lisp :tangle no
dired             ; making dired pretty [functional]
electric          ; smarter, keyword-based electric-indent
;;eww               ; the internet is gross
(ibuffer +icons)           ; interactive buffer management
;;tramp             ; remote files at your arthritic fingertips
(undo +tree)              ; persistent, smarter undo for your inevitable mistakes
vc                ; version-control and Emacs, sitting in a tree
#+end_src

** Term
https://github.com/doomemacs/doomemacs/tree/master/modules/term

#+name:doom-term
#+begin_src emacs-lisp :tangle no
;;eshell            ; the elisp shell that works everywhere
;;shell             ; simple shell REPL for Emacs
;;term              ; basic terminal emulator for Emacs
vterm             ; the best terminal emulation in Emacs
#+end_src

Vterm is great. It does require some additional configuration to work properly. See [[https://github.com/doomemacs/doomemacs/tree/master/modules/term/vterm][relevant doc]] for more.

** Checker
https://github.com/doomemacs/doomemacs/tree/master/modules/checker

#+name:doom-checker
#+begin_src emacs-lisp :tangle no
(syntax +childframe +icons +flymake)             ; tasing you for every semicolon you forget
(spell +aspell +everywhere) ; tasing you for misspelling mispelling
grammar           ; tasing grammar mistake every you make
#+end_src

*** Custom Dictionary
#+begin_src emacs-lisp :tangle "config.el" :noweb no-export :comments no
(setq ispell-dictionary "en"
      ispell-personal-dictionary (concat org-directory ".ispell.en.pws"))
#+end_src

** Tools
https://github.com/doomemacs/doomemacs/tree/master/modules/tools

#+name:doom-tools
#+begin_src emacs-lisp :tangle no
ansible
;;biblio            ; Writes a PhD for you (citation needed)
;;collab            ; buffers with friends
(debugger +lsp)          ; FIXME stepping through code, to help you add bugs
;;direnv
;;docker
;;editorconfig      ; let someone else argue about tabs vs spaces
;;ein               ; tame Jupyter notebooks with emacs
(eval +overlay)     ; run code, run (also, repls)
(lookup +dictionary +docset +offline)              ; navigate your code and its documentation
llm               ; Custom; when I said you needed friends, I didn't mean...
(lsp +peek)      ; M-x vscode
(magit +forge)             ; a git porcelain for Emacs
;;make              ; run make tasks from Emacs
;;pass              ; password manager for nerds
;;pdf               ; pdf enhancements
(terraform +lsp)         ; infrastructure as code
;;tmux              ; an API for interacting with tmux
tree-sitter       ; syntax and parsing, sitting in a tree...
;;upload            ; map local to remote projects via ssh/ftp
#+end_src

** Language
https://github.com/doomemacs/doomemacs/tree/master/modules/lang

#+name:doom-lang
#+begin_src emacs-lisp :tangle no
;;ada               ; In strong typing we (blindly) trust
;;agda              ; types of types of types of types...
;;beancount         ; mind the GAAP
;;(cc +lsp)         ; C > C++ == 1
;;clojure           ; java with a lisp
;;common-lisp       ; if you've seen one lisp, you've seen them all
;;coq               ; proofs-as-programs
;;crystal           ; ruby at the speed of c
;;csharp            ; unity, .NET, and mono shenanigans
data              ; config/data formats
;;(dart +flutter)   ; paint ui and not much else
;;dhall
;;elixir            ; erlang done right
;;elm               ; care for a cup of TEA?
emacs-lisp        ; drown in parentheses
;;erlang            ; an elegant language for a more civilized age
;;ess               ; emacs speaks statistics
;;factor
;;faust             ; dsp, but you get to keep your soul
;;fortran           ; in FORTRAN, GOD is REAL (unless declared INTEGER)
;;fsharp            ; ML stands for Microsoft's Language
;;fstar             ; (dependent) types and (monadic) effects and Z3
;;gdscript          ; the language you waited for
(go +lsp +tree-sitter)         ; the hipster dialect
;;(graphql +lsp)    ; Give queries a REST
;;(haskell +lsp)    ; a language that's lazier than I am
;;hy                ; readability of scheme w/ speed of python
;;idris             ; a language you can depend on
(json +lsp +tree-sitter)              ; At least it ain't XML
;;janet             ; Fun fact: Janet is me!
;;(java +lsp)       ; the poster child for carpal tunnel syndrome
(javascript +lsp +tree-sitter)        ; all(hope(abandon(ye(who(enter(here))))))
;;julia             ; a better, faster MATLAB
;;kotlin            ; a better, slicker Java(Script)
;;latex             ; writing papers in Emacs has never been so fun
;;lean              ; for folks with too much to prove
;;ledger            ; be audit you can be
;;lua               ; one-based indices? one-based indices
markdown          ; writing docs for people to ignore
mermaid           ; Custom. Graphviz++
;;nim               ; python + lisp at the speed of c
(nix +tree-sitter +lsp)               ; I hereby declare "nix geht mehr!"
;;ocaml             ; an objective camel
(org +dragndrop +journal +pandoc +pomodoro +pretty +roam)               ; organize your plain life in plain text
;;php               ; perl's insecure younger brother
;;plantuml          ; diagrams for confusing people more
;;graphviz          ; diagrams for confusing yourself even more
;;purescript        ; javascript, but functional
(python +lsp +poetry +pyright +tree-sitter)            ; beautiful is better than ugly
;;qt                ; the 'cutest' gui framework ever
;;racket            ; a DSL for DSLs
;;raku              ; the artist formerly known as perl6
;;rest              ; Emacs as a REST client
;;rst               ; ReST in peace
(ruby +rbenv +lsp +tree-sitting)     ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
;;(rust +lsp)       ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
;;scala             ; java, but good
;;(scheme +guile)   ; a fully conniving family of lisps
(sh +lsp +tree-sitter)                ; she sells {ba,z,fi}sh shells on the C xor
;;sml
;;solidity          ; do you need a blockchain? No.
;;swift             ; who asked for emoji variables?
;;terra             ; Earth and Moon in alignment for performance.
(web +lsp +tree-sitter)               ; the tubes
(yaml +lsp)              ; JSON, but readable
;;zig               ; C, but simpler
#+end_src

There's a lot there there here.

+ =org= :: I have so many tweaks for this it's its own section. See [[*OrgMode][OrgMode]] for more.

** Emails
https://github.com/doomemacs/doomemacs/tree/master/modules/email

#+name:doom-email
#+begin_src emacs-lisp :tangle no
;;(mu4e +org)
;;notmuch
;;(wanderlust +gmail)
#+end_src

** App
https://github.com/doomemacs/doomemacs/tree/master/modules/app

#+name:doom-app
#+begin_src emacs-lisp :tangle no
calendar
;;emms
 ;;everywhere        ; *leave* Emacs!? You must be joking
;;irc               ; how neckbeards socialize
;;(rss +org)        ; emacs as an RSS reader
#+end_src

** Config
https://github.com/doomemacs/doomemacs/tree/master/modules/config

#+name:doom-config
#+begin_src emacs-lisp :tangle no
literate
(default +bindings +smartparens)
#+end_src

* UI Tweaks
** Save Status
The title bar display string will display =♢= if there exist unsaved buffer modifications & =♦= otherwise.

#+begin_src emacs-lisp :tangle "config.d/+ui.el" :noweb no-export :comments no
(setq doom-fallback-buffer-name "Doom"
      +doom-dashboard-name "Doom Dashboard")

(setq frame-title-format
      '(""
        (:eval
         (if (s-contains-p org-roam-directory (or buffer-file-name ""))
             (replace-regexp-in-string
              ".*/[0-9]*-?" "☰ "
              (subst-char-in-string ?_ ?  buffer-file-name))
           "%b"))
        (:eval
         (let ((project-name (projectile-project-name)))
           (unless (string= "-" project-name)
             (format (if (buffer-modified-p)  " ♢ %s" " ♦ %s") project-name))))))
#+end_src

** Fonts
I use the [[https://www.jetbrains.com/lp/mono/][JetBrains Mono]] for my monospaced font & Huerta Tipografica's [[https://www.huertatipografica.com/en/fonts/alegreya-ht-pro][Alegreya]] serif font.

#+begin_src emacs-lisp :tangle "config.d/+ui.el" :noweb no-export :comments no
(setq
      doom-font (font-spec :family "JetBrains Mono" :size 12 :height 1)
      doom-variable-pitch-font (font-spec :family "Alegreya" :height 1.3)
      doom-big-font (font-spec :family "JetBrains Mono" :size 18))
#+end_src

** Window Size
I'm normally running a titling window manager so this is usually less of an issue, but the emacs window on start is often impractically small. Let's set it to something more reasonable.

#+begin_src emacs-lisp :tangle "config.d/+ui.el" :noweb no-export :comments no
(add-to-list 'default-frame-alist '(height . 64))
(add-to-list 'default-frame-alist '(width . 192))
#+end_src

** Better Linter Messaging
[[https://github.com/konrad1977/flyover?][Flyover]] is a more modern overlay for the flymake & flycheck linting systems in Emacs.

 #+CAPTION: Flyover Example
[[./.resources/flyover_sample.png]]

#+CAPTION: Flyover Example Source
#+begin_src python :tangle no :noweb no-export :comments no :results output silent
import sys

def my_function(  a, b ):
    print( 'Adding numbers...' )
    result = a +  b
    print( "Result is",result )

x = myFunction (5,8 )
#+end_src

#+begin_src emacs-lisp :tangle "packages.el" :noweb no-export :comments no :results output silent
(package! flyover)
#+end_src

We want to enable it with hooks for both flycheck & flymake. There is a lot of configuration settings. I'm just going pretty basic with minor tweaks to the colors & tints used.

#+begin_src emacs-lisp :tangle "config.d/+ui.el" :noweb no-export :comments no
(use-package flyover
  :ensure t
  :hook ((flycheck-mode . flyover-mode)
         (flymake-mode . flyover-mode))
  :custom
  (flyover-checkers '(flycheck flymake))
  (flyover-hide-during-completion t)
  ;; UI tweaks
  (flyover-use-theme-colors t)
  (flyover-text-tint 'lighter)
  (flyover-text-tint-percent 70)
  (flyover-icon-tint 'lighter)
  (flyover-icon-tint-percent 50)
  (flyover-icon-background-tint 'darker)
  (flyover-icon-background-tint-percent 30)
  )
#+end_src

* Projectile Project Search
I primarily keep my code either in =$HOME/Code= or nested inside =$HOME/Documents=. Let's have emacs check these directories
#+begin_src emacs-lisp :tangle "config.el" :noweb no-export :comments no
(setq projectile-project-search-path '("~/Code/" ("~/Documents" . 3)))
#+end_src

While the search is quick I don't think there is enough value in making this happening automatically on startup. As such I'm going to leave =projectile-auto-discover= & =projectile-cleanup-known-projects= as =nil=. Instead use the =projectile-discover-projects-in-search-path= to add projects & =projectile-cleanup-known-projects= to cleanup removed projects.

This part isn't needed but let's set these values as =nil=.
#+begin_src emacs-lisp :tangle "config.el" :noweb no-export :comments no
(setq projectile-auto-discover nil)
(setq projectile-cleanup-known-projects nil)
#+end_src

Further documentation on this can be found [[https://docs.projectile.mx/projectile/usage.html#automated-project-discovery][here]].

* Keybinds
I use Evil for most things, & most of Doom's defaults. I'm going to try and centralise keybinds in to the =$DOOMDIR/config.d/+keybinds.el=. This isn't always going to happen as the =use-package!= can be really helpful for setting up.

Doom ships with a =map!= macro (see [[https://github.com/doomemacs/doomemacs/blob/master/docs/faq.org#bind-my-own-keys-or-change-existing-ones][here]] for why).  Documentation for this can be found [[https://github.com/doomemacs/doomemacs/blob/master/docs/faq.org#bind-my-own-keys-or-change-existing-ones][here]], but in broad strokes:
1. Start with a =map!= macro.  You can use
2. When possible use =:when= to limit scope.
3. Use =:prefix= and =:prefix-map= where possible.
4. Always use =:desc= to describe what is being done.

Below is a sample snippet from Rameez Khan's [[https://rameezkhan.me/posts/2020/2020-07-03--adding-keybindings-to-doom-emacs/][blog]].
   #+begin_src emacs-lisp :tangle no
(map! :leader
      (:prefix-map ("a" . "applications")
                   (:prefix ("j" . "journal")
                    :desc "New journal entry" "j" #'org-journal-new-entry
                    :desc "Search journal entry" "s" #'org-journal-search)))
   #+end_src

** =whichkey= response time
By default which key has a delay of 1 second before it responds. Lets make it a little more responsive by halving that.

#+begin_src emacs-lisp :tangle "config.d/+keybinds.el" :noweb no-export :comments no
(setq which-key-idle-delay 0.5)
#+end_src

** Icons
I make liberal use of the =nerd-icons-insert= function. To make it a little more accessible I'm binding it to ~SPC i i~.

#+begin_src emacs-lisp :tangle "config.d/+keybinds.el" :noweb no-export :comments no
(map! :leader
      (:prefix "i"
       :desc "icons" "i" #'nerd-icons-insert
       )
      )
#+end_src

Its a little annoying to have to go through the leader key for this all the time. So lets also bind it to ~M-I~ (~SHIFT-META-i~) in insert mode.

#+begin_src emacs-lisp :tangle "config.d/+keybinds.el" :noweb no-export :comments no :results silent
(map! :i "M-I" #'nerd-icons-insert)
#+end_src

* OrgMode
OrgMode is ridiculous. It's a combination note taking, todo-list, jupyter notebook, & calendaring major-mode for Emacs. It is incredibly configurable and extendable, and feature rich to the degree that discovering spreadsheet support after using it for 6 months seems to be a rite of passage. It's wonderful and terrible in a lot of the ways emacs itself is. That said its one of the primary reasons I can't leave emacs.

I currently have my files synced via git (see [[*Syncing][Syncing]] for details). But I've yet to find a good mobile clients. I've bounced off [[https://logseq.com/][LogSeq]] multiple times. [[https://xenodium.com][xenodium's]] great [[https://xenodium.com/plain-org-for-ios][PlainOrg]], doesn't vibe well with a roam style setup. I'm tempted to build my own implementation ([[https://xkcd.com/927/][relevant XKCD]]), but given the small child & a growing list household projects, this isn't likely soon.

** Making Scripting a little easier.
Doom is pretty quick, but we don't need the entire thing for scripts. The =$DOOMDIR/config.d/+org.el= should be written in such a way to allow it to be imported for easy scripting. To wit

#+begin_src emacs-lisp :tangle "config.d/+org.el" :noweb no-export :comments no
(require 'org)
#+end_src

** Directory Structure
For all of my machines if OrgMode is used then my org notes are stores at =$HOME/org=. Everything is then synced to GitHub.

#+begin_src emacs-lisp :tangle "config.d/+org.el" :noweb no-export :comments no
(custom-set-variables '(org-directory "~/org/"))
#+end_src

I've tried to segment things to keep them clean. But entropy gonna entropy & some things either didn't work as well as I had hoped or proved squishy and escaped containment. As for writing this is what my notes repo looks like.

#+begin_src bash :tangle no :exports result :results output verbatim
cd ${HOME}/org && tree -a -L 1 --gitignore --dirsfirst | grep -v "\.git"
#+end_src

#+RESULTS:
#+begin_example
.
├── .archive
├── .attach
├── journals
├── logseq
├── pages
├── projects
├── roam
├── work
├── .dir-locals.el
├── .ispell.en.pws
├── README.org
└── calendar.org

10 directories, 5 files
#+end_example

So lets go over the list.
+ =.archive= :: Graveyard for old notes.
+ =.attach= :: attachments for notes. Images, json, etc.
+ =logseq= :: I took a stab at logseq. I'm still stabbing but its not going so well.
+ =pages= :: A [[https://en.wikipedia.org/wiki/Zettelkasten][zettelkasten]] style collection of notes using [[https://github.com/org-roam/org-roam][org-roam]].
+ =projects= :: A defunct idea to symlink in todo lists for projects.
+ =work= :: Post mortems, white papers, design docs, and a very heavily used =todo.org= that I use to track my current work & billable hours for capitalization.
+ =journals= :: Daily lab style note books. This is my default catch all. Everything from call details with my local mechanic to implementation log. These link to =pages= using org-roam, and fully leverage the backlink mechanism.

Then we have the loose files.
+ =.dir-locals.el= :: Directory specific variables for elisp. See [[https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Variables.html][here]] for more.
+ =.ispell.en.pws= :: Personal dictionary for everything.
+ =README.org= :: GitHub readme detailing folder structure in more detail than here.
+ =calendar.org= :: Autosync calendar using [[https://vdirsyncer.pimutils.org/en/stable/][vdirsyncer]]  [[https://github.com/pimutils/khal][khal]]  [[https://github.com/emacsmirror/khalel][khalel]]. See [[*Calendaring][Calendaring]] for more.

#+begin_src emacs-lisp :tangle "config.d/+org.el" :noweb no-export :comments no
(setq org-download-image-dir (concat org-directory ".attach/"))
#+end_src

the zettelkasten pages directory.
** Work Directory
As mentioned previously this folder contains of work related notes & documents, but the workhorse of this section is the =work/todo.org= file. To make my life easier lets lower the barrier for getting things into my this file.

First let's setup a variable for this folder

#+begin_src emacs-lisp :tangle "config.d/+org.el" :noweb no-export :comments no
(defvar org-directory-work (concat org-directory "work/") "Location for work subdirectory.")
#+end_src

All entries should be added to the "Inbox" tree. Lets add a new =org-capture= to put these there.

#+begin_src emacs-lisp :tangle "config.d/+org.el" :noweb no-export :comments no
(defvar-local +org-capture-work-todo-file
    (expand-file-name "todo.org" org-directory-work))

emacs/config.el
;;; -*- lexical-binding: t; -*-

;;; This file is generated via tangles from the config.org file. Do not modify this file.
;;; Any modifications here will be clobbered and versioned over. Seriously just don't.

;;; -*- lexical-binding: t; -*-
;;; This file is generated via tangles from the config.org file. Do not modify this file.
;;; Any modifications here will be clobbered and versioned over. Seriously just don't.

(setq fancy-splash-image (concat doom-user-dir "./.resources/doom.png"))

(setq-hook! 'sql-mode-hook +format-with 'pgformatter)

(setq ispell-dictionary "en"
      ispell-personal-dictionary (concat org-directory ".ispell.en.pws"))

(setq doom-fallback-buffer-name "Doom"
      +doom-dashboard-name "Doom Dashboard")

(setq frame-title-format
      '(""
        (:eval
         (if (s-contains-p org-roam-directory (or buffer-file-name ""))
             (replace-regexp-in-string
              ".*/[0-9]*-?" "☰ "
              (subst-char-in-string ?_ ?  buffer-file-name))
           "%b"))
        (:eval
         (let ((project-name (projectile-project-name)))
           (unless (string= "-" project-name)
             (format (if (buffer-modified-p)  " ♢ %s" " ♦ %s") project-name))))))

(setq doom-font (font-spec :family "Hasklig" :size 14 :height 1)
      doom-variable-pitch-font (font-spec :family "Alegreya" :height 1.3)
      doom-big-font (font-spec :family "Hasklig" :size 18))

(load! "+keybinds")

(setq which-key-idle-delay 0.1)

(add-to-list 'auto-mode-alist '("/\\(tmp\\|private/var\\)/.*/tmp_.*\\.txt" . markdown-mode))

(load! "+orgmode")

(use-package! mermaid-ts-mode
  :defer t)

(use-package! mermaid-mode
  :config
  (setq mermaid-mmdc-location "/opt/homebrew/bin/mmdc")
  (setq mermaid-flags "--iconPacks @iconify-json/logos")
  :mode "\\.mmd\\'"
  )

(use-package! copilot
  :hook (prog-mode . copilot-mode)
  :bind (:map copilot-completion-map
              ("[tab]" . 'copilot-accept-completion)
              ("[M-tab]" . 'copilot-accept-completion)
              ("M-n" . 'copilot-accept-completion) ; Fallback should tab act weird
              ))

(use-package! copilot-chat
  :config
  (setq copilot-chat-default-model "claude-sonnet-4")
  )
